src/rdb.c

```diff
@@ -3070,9 +3082,24 @@ int rdbLoadRioWithLoadingCtx(rio *rdb, int rdbflags, rdbSaveInfo *rsi, rdbLoadin
                 goto eoferr;
             if ((expires_size = rdbLoadLen(rdb,NULL)) == RDB_LENERR)
                 goto eoferr;
-            expandDb(db, db_size);
+            /* In cluster mode we don't resize whole DB, instead we rely on RDB_OPCODE_SLOT_INFO for resizing slots. */
+            if (!server.cluster_enabled) {
+                expandDb(db, db_size);
+            }
```

Comment:
if we read an old RDB, it may still be beneficial not to ignore this opcode and assume even distribution.

Reply:
Done. One challenge was how to avoid resizing DB if slot info was present (since db size OPCODE comes before slot info, but it should be no-op if slot info has more specific instructions). To achieve this, I moved DB resizing after OPCODE loading phase.